### PR TITLE
Simplify AutoAugment graph

### PR DESCRIPTION
**Pull request by stiepan**
## Category:
**Refactoring** (*Redesign of existing code that doesn't affect functionality*

## Description:
This PR combines a couple of improvements to AutoAugment processing graph.

1. The list of augmentations to split processing into is created independently for each stage - this can reduces the size of split-merge tree generated by the `select` operation.

For example, the default v0 policy uses 25 sub-policies, each sub-policy is a sequence of (at most) 2 augmentations. Even though there are 25 sub-policies, there are only 11 unique augmentations present in those sub-polcies. DALI already splits the computation only in 11 parts. However, it can be further improved if the operators used in given stages of sub-policies form even smaller groups. In case of the default v0 policy, it is 8 vs 11.

In other words, when the comutation is split to apply the "i-th" augmentation in a sequence according to selected sub-polciy, there is no need to take into account all augmentations present in all sub-polcies, but only the augmentations at the i-th position in all sub-policies.

2. The `signed_bin` helper that handles random negation of magnitudes can now accept additional shape parameter. With that, if you have a multiple random augmentations to apply in a sequence, you can generate the random signs of magnitudes once for all stages, rather than do it for every stage separately.

### Affected modules and functionalities:
AutoAugment (points 1., 2.), RandAugment (point 2.).

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3357

### CODE REVIEW

File: dali/python/nvidia/dali/auto_aug/core/_augmentation.py [hunk 1]
```diff
@@ -45,19 +45,43 @@ class _UndefinedParam:
...
 
-    def __init__(self, magnitude_bin: Union[int, _DataNode],
-                 random_sign: Optional[_DataNode] = None, seed=None):
+    def __init__(self, magnitude_bin: Union[int, _DataNode], random_sign: _DataNode,
+                 signed_magnitude_idx: _DataNode):
+        self._magnitude_bin = magnitude_bin
+        self._random_sign = random_sign
+        self._signed_magnitude_idx = signed_magnitude_idx
+
+    def __getitem__(self, idx: int):
```
Comment (stiepan) [inline dali/python/nvidia/dali/auto_aug/core/_augmentation.py:54]: This enables the `signed_bin(mutli_stage_bin)[stage_idx]`.

**Review by jantonguirao — APPROVED** [review]

**Comment by stiepan** [comment]
!build

File: dali/python/nvidia/dali/auto_aug/auto_augment.py [hunk 2]
```diff
@@ -111,16 +111,14 @@ def apply_auto_augment(policy: Policy, sample: _DataNode, seed: Optional[int] =
...
-        if use_signed_magnitudes:
-            magnitude_bin = signed_bin(magnitude_bin)
         if should_run[stage_id] < run_probabilities[stage_id]:
-            op_kwargs = dict(sample=sample, magnitude_bin=magnitude_bin,
+            op_kwargs = dict(sample=sample, magnitude_bin=magnitude_bins[stage_id],
                              num_magnitude_bins=policy.num_magnitude_bins, **kwargs)
-            sample = _pretty_select(augmentations, aug_ids[stage_id], op_kwargs,
+            sample = _pretty_select(augmentations[stage_id], aug_ids[stage_id], op_kwargs,
                                     auto_aug_name='apply_auto_augment',
                                     ref_suite_name='get_image_net_policy')
```
Comment (klecki) [inline dali/python/nvidia/dali/auto_aug/auto_augment.py:123]: Honestly, this is so cool ;)

**Review by klecki — APPROVED** [review]

File: dali/python/nvidia/dali/auto_aug/auto_augment.py [hunk 3]
```diff
@@ -223,15 +221,30 @@ def _sub_policy_to_magnitude_bin_map(policy: Policy) -> _DataNode:
     return types.Constant(magnitude_bin)
 
 
-def _sub_policy_to_augmentation_map(policy: Policy) -> Tuple[_DataNode, List[_Augmentation]]:
+def _sub_policy_to_augmentation_map(policy: Policy) -> Tuple[_DataNode, List[List[_Augmentation]]]:
+    """
+    Creates a matrix of operators to be called for given sub policy at given stage.
+    The output is a tuple of matrix `m` and per stage operators augmentations list `augments`,
+    such that for policy `sub_policy_idx` as the `stage_idx`-ith operation in a sequence, the
+    `augments[stage_idx][m[sub_policy_idx][stage_idx]]` operator should be called.
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/python/nvidia/dali/auto_aug/auto_augment.py:229]: Sure, I'll merge this one, revase the https://github.com/NVIDIA/DALI/pull/4753 and expand the comment there.

**Pull request merged**